### DISCUSSION TITLE
replacing deprecated SSL_CTX_set_tmp_ecdh

### DISCUSSION THREAD

**Discussion by rdwr-vitalyk** [openssl/openssl#23070] [Q&A] [answered]
hi,
 i try to remove deprecated functions from our program while we we upgrading to openssl 3.

 i have this code that modifies ssl ctx structure:
```
{
 EC_GROUP  *value = PEM_read_bio_ECPKParameters(mem, NULL, password_callback, NULL);
 EC_KEY *ec_key = EC_KEY_new();
 EC_KEY_set_group(ec_key, value) ;
 SSL_CTX_set_tmp_ecdh(ctx, ec_key)
}
```
 can you please advise  how i can do the same without using deprecated calls?
thank you!

**Comment by t8m** [comment, marked as answer]
First of all I would recommend not calling SSL_CTX_set_tmp_ecdh at all. I.e., let OpenSSL manage the ECDH group selection itself.

However you can use SSL_set1_groups_list() to set the groups supported - if you set just one group it will be equivalent to the SSL_CTX_set_tmp_ecdh() call. You can get the curve name from an EC_GROUP by using EC_GROUP_get_curve_name() and OSSL_EC_curve_nid2name().

**Comment by rdwr-vitalyk** [comment]
Thank you!
I have one more question...
Regarding  EC_GROUP, we used PEM_read_bio_ECPKParameters to load it.
But it is deprecated.

I tried this:
    // TYPE *PEM_read_bio_TYPE(BIO *bp, TYPE **a, pem_password_cb *cb, void *u);
    EC_GROUP *ecg = nullptr;
    PEM_read_bio_EC_GROUP(mem, &ecg, nullptr, nullptr);

But PEM_read_bio_EC_GROUP is absent.
Can you please advise how we can load EC_GROUP from BIO using non-deprecated functions ?


Thanks!
Vitaly

**Reply by t8m** [reply]
Could you read the PEM file as a EVP_PKEY? Use PEM_read_bio_Parameters_ex() or PEM_read_bio_Parameters(). You can get the group name from the EVP_PKEY with EVP_PKEY_get_utf8_string_param() then.

**Comment by rdwr-vitalyk** [comment]
Thank you very much!

Vitaly
